### PR TITLE
fix/Refactored codebase

### PR DESCRIPTION
**Pull request by OnicaAdrianCatalin**
- Placed the hardcoded strings to resource file strings,
-  changed the authentication logic separated it in Services classes
-  fixed Trashtype enum and changed timePeriod from list to enum
- fixed CustomTextFields adding a custom value to non-related password fields to VisualTrasformation.None

### CODE REVIEW

File: app/src/main/java/com/example/trashminder/di/AppModules.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.trashminder.presentation.auth.signup.SignUpViewModel
+import com.example.trashminder.services.FirebaseAuthService
+import com.example.trashminder.services.FirebaseAuthServiceImpl
+import org.koin.androidx.viewmodel.dsl.viewModel
+import org.koin.dsl.module
+
+val appModule = module {
+    single<FirebaseAuthService> { FirebaseAuthServiceImpl() }
+    viewModel {
+        LoginViewModel(get())
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/di/AppModules.kt]: Would be nice to have named parameter here
E.g: ` LoginViewModel(authService = get())`

File: app/src/main/java/com/example/trashminder/model/Reminder.kt [hunk 2]
```diff
@@ -1,8 +1,11 @@
 package com.example.trashminder.model
 
+import com.example.trashminder.utils.TimePeriod
+import com.example.trashminder.utils.TrashType
+
 data class Reminder(
     val id: Int = 0,
-    val type: String = "",
+    val type: TrashType = TrashType.Residuals,
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/model/Reminder.kt]: Do we need these default values or could we remove them?

File: app/src/main/java/com/example/trashminder/presentation/auth/login/LoginFragment.kt [hunk 3]
```diff
@@ -77,7 +77,7 @@ class LoginFragment : Fragment() {
     fun LoginScreen() {
         val email = remember { mutableStateOf("") }
         val password = remember { mutableStateOf("") }
-        val viewModel = viewModel<LoginViewModel>()
+        val viewModel = get<LoginViewModel>()
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/auth/login/LoginFragment.kt:80]: Why the change from `viewModel -> get`? Is it because of the Koin injection?

File: app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt [hunk 4]
```diff
@@ -20,38 +24,23 @@ class SignUpViewModel : ViewModel() {
...
-                    .addOnCompleteListener { task ->
-                        if (task.isSuccessful) {
-                            _authResult.value = Result.Success(Unit)
-                        } else {
-                            if (task.exception is FirebaseAuthUserCollisionException) {
-                                _authResult.value =
-                                    Result.Failure(AuthErrors.EMAIL_ALREADY_IN_USE)
-                            }
+        viewModelScope.launch {
+            if (assertFieldsNotEmpty(email, password, confirmPassword, firstName, lastName)) {
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt]: You should try avoid nesting if statements if possible. For such scenarios you can use the "early return pattern" (feel free to look it up and learn about it). This means that we check the opposite of the condition we need to execute some code and returning from the function if the conditions are not met. So in this case this would look something like this:

```
            if (!assertFieldsNotEmpty(email, password, confirmPassword, firstName, lastName)) {
                _authResult.value = Result.Failure(AuthErrors.EMPTY_FIELDS)
                return@launch
            }

            if (confirmPassword != password) {
                _authResult.value = Result.Failure(AuthErrors.PASSWORD_MISMATCH)
                return@launch
            }

            authService.register(email, password).onSuccess {
                _authResult.value = Result.Success(Unit)
            }.onFailure {
                if (it is FirebaseAuthUserCollisionException) {
                    _authResult.value = Result.Failure(AuthErrors.EMAIL_ALREADY_IN_USE)
                }
            }
```

Comment (iugaliviu) [reply app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt]: Also, I notice you only handle the email already being used failure. What about other cases? Server not responding or internet connection not being available?

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt [hunk 5]
```diff
@@ -29,57 +29,59 @@ import com.example.trashminder.presentation.theme.blueTrash
 import com.example.trashminder.presentation.theme.greenTrash
 import com.example.trashminder.presentation.theme.red
 import com.example.trashminder.presentation.theme.yellow
-import com.example.trashminder.utils.TrashTypeItems
+import com.example.trashminder.utils.TrashType
+import com.example.trashminder.utils.toLocalizedString
 
 @Composable
-fun CustomItem(reminder: Reminder) {
+fun CustomItem(reminder: Reminder?, context: Context) {
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: Why is the reminder nullable?

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt [hunk 6]
```diff
@@ -29,57 +29,59 @@ import com.example.trashminder.presentation.theme.blueTrash
...
-        TrashTypeItems.Menajer.trash -> CardItem(
+        TrashType.Residuals -> CardItem(
+            context = context,
             reminder = reminder,
             startColor = black,
-            type = TrashTypeItems.Menajer.type,
             painterId = R.drawable.report_feat_removebg_preview,
             modifierImage = Modifier.size(110.dp, 90.dp)
         )
+        else -> {}
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: I think we can remove this

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt [hunk 7]
```diff
@@ -29,57 +29,59 @@ import com.example.trashminder.presentation.theme.blueTrash
...
+        TrashType.Plastic -> CardItem(
+            context = context,
             reminder = reminder,
             startColor = yellow,
-            type = TrashTypeItems.Plastic.type,
             painterId = R.drawable.plastic_bottle_png_download_image,
             modifierImage = Modifier.size(90.dp)
         )
-        TrashTypeItems.Metale.trash -> CardItem(
+        TrashType.Metals -> CardItem(
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: This should be singular as in `Metal`

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt [hunk 8]
```diff
@@ -29,57 +29,59 @@ import com.example.trashminder.presentation.theme.blueTrash
...
+        TrashType.Glass -> CardItem(
+            context = context,
             reminder = reminder,
             startColor = greenTrash,
-            type = TrashTypeItems.Sticla.type,
             painterId = R.drawable.broken_bottle_png23,
             modifierImage = Modifier.size(130.dp, 80.dp)
         )
-        TrashTypeItems.Menajer.trash -> CardItem(
+        TrashType.Residuals -> CardItem(
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: Also singular -> `Residual`

File: app/src/main/java/com/example/trashminder/utils/TrashType.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package com.example.trashminder.utils
+
+import android.content.Context
+import com.example.trashminder.R
+
+enum class TrashType {
+    Plastic,
+    Metals,
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/utils/TrashType.kt]: Also singular `Metal`

File: app/src/main/res/values-ro-rRO/strings.xml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="app_name">Trashminder</string>
+    <string name="menu_home_fragment">Home</string>
+    <string name="menu_calendar_fragment">Calendar</string>
+    <string name="menu_settings_fragment">Settings</string>
+    <string name="fields_cannot_be_not_empty_validation">Fieldurile nu pot fi necompletate</string>
+    <string name="email_or_password_incorrect_validation">Emailul sau parola sunt incorece</string>
```
Comment (iugaliviu) [inline app/src/main/res/values-ro-rRO/strings.xml]: incorece? :)

File: app/src/main/res/values-ro-rRO/strings.xml [hunk 11]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="app_name">Trashminder</string>
+    <string name="menu_home_fragment">Home</string>
+    <string name="menu_calendar_fragment">Calendar</string>
+    <string name="menu_settings_fragment">Settings</string>
+    <string name="fields_cannot_be_not_empty_validation">Fieldurile nu pot fi necompletate</string>
```
Comment (iugaliviu) [inline app/src/main/res/values-ro-rRO/strings.xml]: Câmpurile*

File: app/src/main/res/values-ro-rRO/strings.xml [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    <string name="email_already_exists_validation">Emailul deja exista</string>
+    <string name="calendar_text">CALENDAR</string>
+    <string name="add_button_content_description">Add button</string>
+    <string name="log_out_button">Iesi din cont</string>
+    <string name="choose_trash_type">Alege tipul de gunoi</string>
+    <string name="save_button">Salveaza</string>
+    <string name="repetition">Repetare</string>
+    <string name="date_and_time_button">Data si ora</string>
+    <string name="trash_type_paper">Hartie</string>
+    <string name="trash_type_metals">Metale</string>
```
Comment (iugaliviu) [inline app/src/main/res/values-ro-rRO/strings.xml]: Singular :)

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt (hunk 5 above)
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: Found a better way of handling strings. This context can be removed

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt [hunk 13]
```diff
@@ -93,37 +95,24 @@ private fun CardItem(
             ) {
                 Column(modifier = Modifier.padding(10.dp)) {
                     Text(
-                        text = type,
+                        text = reminder.type.toLocalizedString(context),
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: A better way of handling this is using `stringResource`. Apparently this function is available in every @Composable function and you can use it to get a string. This is based on the assumption that a composable always has a context since it is a view so it is safe to fetch the string. This would then become 
`stringResource(reminder.type.toResourceId())`. Where `toResourceId` is the old `toLocalizedString(context)`

File: app/src/main/java/com/example/trashminder/utils/TrashType.kt [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    Paper,
+    Glass,
+    Residuals
+}
+
+fun TrashType.toLocalizedString(context: Context): String {
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/utils/TrashType.kt]: This can be renamed `toResourceId():Int` and make it return just the resource id

File: app/src/main/java/com/example/trashminder/utils/TrashType.kt [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    return when (this) {
+        TrashType.Paper -> context.getString(R.string.trash_type_paper)
+        TrashType.Metals -> context.getString(R.string.trash_type_metals)
+        TrashType.Glass -> context.getString(R.string.trash_type_glass)
+        TrashType.Residuals -> context.getString(R.string.trash_type_residuals)
+        TrashType.Plastic -> context.getString(R.string.trash_type_plastic)
+    }
+}
+
+fun String.toTrashType(context: Context): TrashType? {
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/utils/TrashType.kt]: This should not exist. We should operate with the trash type alone without converting it back and forth from string to type.  `val trashType = remember { mutableStateOf("") }` in `NewReminderScreen` should be a  `TrashType` not a `String`

**Review by iugaliviu — CHANGES_REQUESTED** [review]

File: app/src/main/res/values-ro-rRO/strings.xml (hunk 12 above)
Comment (iugaliviu) [reply app/src/main/res/values-ro-rRO/strings.xml]: @OnicaAdrianCatalin English still has the plural. Also the string naming has plural. Please fix this as well

File: app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt (hunk 4 above)
Comment (iugaliviu) [reply app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt]: @OnicaAdrianCatalin Now that you have all the error handling, please create a separate function for this (E.g: `private fun handleError(exception: FirebaseException)`. Feel free to use a when statement when deciding what value should `_authResult` be assigned. Usually "if" statements work ok for 1-2 checks; when we start checking for 3 or more situations, a when statement is more handy if the check is made on the same type

File: app/src/main/java/com/example/trashminder/di/AppModules.kt (hunk 1 above)
Comment (iugaliviu) [reply app/src/main/java/com/example/trashminder/di/AppModules.kt]: @OnicaAdrianCatalin This comment appears to be outdated, but code is basically the same as before. Did you make any changes here?

File: app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt (hunk 4 above)
Comment (iugaliviu) [reply app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt]: Also, `!assertFieldsNotEmpty`, this is a double negative so maybe `assertFieldsEmpty` would be better suited? It is always easier to read `fields empty` than `fields are not not empty`

File: app/src/main/java/com/example/trashminder/model/Reminder.kt (hunk 2 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/java/com/example/trashminder/model/Reminder.kt]: We need to keep these default values to avoid RuntimeException

File: app/src/main/java/com/example/trashminder/presentation/auth/login/LoginFragment.kt (hunk 3 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/java/com/example/trashminder/presentation/auth/login/LoginFragment.kt:80]: Yes! Koin is forcing me to use get instead viewModel.

File: app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt (hunk 5 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/java/com/example/trashminder/presentation/createdReminder/CustomItem.kt]: I think i forgot to remove that "?". Thanks for noticing that mistake. :)

File: app/src/main/res/values-ro-rRO/strings.xml (hunk 12 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/res/values-ro-rRO/strings.xml]: Fixed

File: app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt (hunk 4 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/java/com/example/trashminder/presentation/auth/signup/SignUpViewModel.kt]: > Also, `!assertFieldsNotEmpty`, this is a double negative so maybe `assertFieldsEmpty` would be better suited? It is always easier to read `fields empty` than `fields are not not empty`

I also believe that assertFieldsEmpty would be better suited in this case. Thanks for advice. :)

File: app/src/main/java/com/example/trashminder/di/AppModules.kt (hunk 1 above)
Comment (OnicaAdrianCatalin) [reply app/src/main/java/com/example/trashminder/di/AppModules.kt]: I'm sorry i misunderstood the comment. I thought you said to add FirebaseAuthServiceImpl parameter instead of get(). Now is all good.

File: app/src/main/java/com/example/trashminder/presentation/auth/login/LoginViewModel.kt [hunk 16]
```diff
@@ -1,48 +1,48 @@
...
                 }
-        } else {
-            _authResult.value = Result.Failure(AuthErrors.EMPTY_FIELDS)
+            }
         }
-    }
 
-    companion object {
-        private const val TAG = "LoginViewModel"
+    private fun handleError(exception: FirebaseException) {
```
Comment (iugaliviu) [inline app/src/main/java/com/example/trashminder/presentation/auth/login/LoginViewModel.kt:38]: ❤️

**Review by iugaliviu — APPROVED** [review]
Splendid! Good job!
